### ISSUE TITLE
Using Istio with CronJobs

### ISSUE THREAD

**Issue by Stono** [istio/istio#11659]
Hey all, 
I have an issue with Istio when used in conjunction with CronJobs or Jobs, in that when the primary pod completes, the "Job" never completes because istio-proxy is still running:

```
NAME                                  READY     STATUS    RESTARTS   AGE
backup-at-uk-1549872000-7hrx7         1/2       Running   0          34m
```

I tried adding the following to the end of the primary pod script as suggested by @costinm in https://github.com/istio/istio/issues/6324, but that doesn't work (envoy exits, proxy doesn't):
```
curl --max-time 2 -s -f -XPOST http://127.0.0.1:15000/quitquitquit
OK
```

Which seems to cause envoy to exit correctly, however the istio-proxy process is still running:
```
istio-proxy@backup-at-uk-1549872000-7hrx7:/$ ps aux
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
istio-p+       1  0.0  0.0  32640 18820 ?        Ssl  08:00   0:00 /usr/local/bin/pilot-agent proxy sidecar --concurrency 1 --configPath /etc/istio/proxy --binaryPath /usr/local/bin/envoy --serviceCluster helm-solr-backup --drainDuration
```

Despite it no longer listening:
```
istio-proxy@backup-at-uk-1549872000-7hrx7:/$ netstat -plunt
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
```

The main pod can't send a SIGTERM to istio-proxy because it doesn't have permission to do so (quite rightly) so I'm a little stuck.

The only hacky thing I can think of doing is adding a readinessProbe to istio-proxy which checks to see if it's listening and if it isn't, sends the SIGTERM.

Thoughts?

**Comment by Stono** [comment]
Associated: https://github.com/kubernetes/kubernetes/issues/25908

**Comment by huikang** [comment]
Same issue for me.

**Comment by Stono** [comment]
For those who are interested, we worked around this by adding a livenessProbe to the sidecar injector for istio-proxy:

```
    livenessProbe:
      exec:
        command:
          - /usr/local/bin/liveness.sh
      initialDelaySeconds: 3
      periodSeconds: 10
      failureThreshold: 5
```

And then the script looks like this:
```
#!/bin/bash
set -e 
if ! pidof envoy &>/dev/null; then
  if pidof pilot-agent &>/dev/null; then
    echo "Envoy is not running, exiting istio-proxy"
    kill -s TERM $(pidof pilot-agent)
    exit 1 
  fi 
fi

if ! netstat -plunt | grep 15001; then 
  echo "Istio-proxy not listening on 15001"
  exit 1
fi 
exit 0

```

**Comment by huikang** [comment]
@Stono thanks for sharing your work around. Are the above script and the livenessProbe added to the cronjob yams file? I am asking because I could not understand how to `adding a livenessProbe to the sidecar injector for istio-proxy`. Thanks.

**Comment by Stono** [comment]
I use the auto sidecar injector.  The probes are added to the injected
template for the istio-proxy pod.

Karl

**Comment by huikang** [comment]
Hi, @Stono, it is still unclear to me how the liveness probe can be added to the istio-proxy pod (my understanding is that the istio-proxy image is not managed by the end user).

Could you point me to some online resources? Thanks.

**Comment by Stono** [comment]
Ohhh I get you.
So in our CI pipeline we build our own istio-proxy image which pulls from
the main image and copies these files in.

Another thing you could do is mount them from a ConfigMap.

Karl

**Comment by liamawhite** [comment]
I would imagine native Istio support for this would require a similar approach. Exposing an additional path on the pilot-agent server that tells it to shut down. This would still require the batch job calls that endpoint after it is finished.

I wonder if there is something else we could hook into in Kube that would automatically call this endpoint once the job is finished?

**Comment by Stono** [comment]
@liamawhite do you know of any way to make istio-proxy exit with a 0 status code, at the moment if i send a `SIGTERM` I get 137, which causes a cronjob failure.  I need an exit signal which will shutdown with a 0

**Comment by Bessonov** [comment]
Similar issue: https://github.com/istio/istio/issues/11045

**Comment by liamawhite** [comment]
I think the lastest in release-1.1 should return 0. I will try to find some time to verify.

**Comment by Stono** [comment]
Seems to be fine for me in 1.1.1

**Comment by Stono** [comment]
We have started doing this recently btw folks:
```
            command: ["/bin/bash", "-c"]
            args:
              - |
                trap "curl --max-time 2 -s -f -XPOST http://127.0.0.1:15000/quitquitquit" EXIT
                while ! curl -s -f http://127.0.0.1:15020/healthz/ready; do sleep 1; done
                sleep 2
                {{ $.Values.cron.command }}
```

This will:
  - Wait for proxy to be ready
  - Tell it to quit when done

**Comment by mikesimons** [comment]
@stono Have you observed the sidecar living for a while after receiving quitquitquit? Ours are living for another minute or so before exiting (although 15000 gets closed immediately).

We're also getting log spam on Completed jobs until we delete them:
```
info	Envoy proxy is NOT ready: failed retrieving Envoy stats: Get http://127.0.0.1:15000/stats?usedonly: dial tcp 127.0.0.1:15000: connect: connection refused
```

**Comment by Stono** [comment]
Yeah so we also have a liveness probe that looks for 15000 listening and if
it's missing exits.

It's hacky.

**Comment by Bessonov** [comment]
Activity!

**Comment by howardjohn** [comment]
In 1.3 we added a new `/quitquitquit` endpoint to pilot agent which should resolve this. Its not perfect since it requires some manual action, but I think a longer term solution would depend on https://github.com/kubernetes/kubernetes/issues/65502 or https://github.com/istio/istio/issues/11366 or similar. If there is anything else we can do to improve these in the short term I would be happy to take a look at

**Comment by drshade** [comment]
Took me a while to figure this out, so wanted to help any others that are battling with this. My approach was to copy @Stono above, but change the port number to 15020 (to hit the pilot agent, not the envoy proxy directly). This is only available in Istio 1.3 onwards. 

```
command: ["/bin/bash", "-c"]
args:
 - |
   trap "curl --max-time 2 -s -f -XPOST http://127.0.0.1:15020/quitquitquit" EXIT
   while ! curl -s -f http://127.0.0.1:15020/healthz/ready; do sleep 1; done
   echo "Ready!"
   < your job >
```

Hope that helps!

**Comment by smawson** [comment]
Reopening this as I'd like to see if we can put together a short-term fix while waiting for sidecar containers / CRI since the workaround of making changes to every Cron job is not great for teams with lots of Cron jobs. 

The shorter term and hackier the better :)

**Comment by sjmiller609** [comment]
We are making use of the sidecar auto-injector. I am missing something here - how do I configure the preStop hook for the istio-proxy container? Do I modify the istio-sidecar-injector configmap? or is there a way to provide this configuration to istio (I have not found that)?
Is this what people are doing?

1) install istio (e.g. helm chart)
2) modify configmap istio-sidecar-injector (drifting from the helm configuration)
3) install the rest of your services

then if you update istio helm chart for any other reason, re-apply step 2?

**Comment by sjmiller609** [comment]
The answer to my question is to use this setting:
.Values.global.proxy.lifecycle

available since 1.4.3

**Comment by stewartbutler** [comment]
I spoke with @smawson and we bounced a few ideas around. Here's a summary of what we talked about. 

@smawson, if I missed something please let me know -- got distracted before I could finish flushing my memory to disk.

**Options:**
---

**External Detection Controller**
This option is a controller similar to in https://github.com/istio/cni/pull/226.

We create a new controller that will serve TLS connections on a kubernetes service port. This controller will, on request, query the API server and check if the workload of a cronjob pod with an istio sidecar has entered a terminal status, and return the result to the caller.

We add new code either to the proxy or add a custom `init` (similar to https://github.com/krallin/tini or https://github.com/Yelp/dumb-init) which will reach out to the controller and send a command to cleanly shut down the proxy when the controller reports that all non-proxy containers in its pod have exited.

**Benefits**
- Controller can be integrated into future versions of Pilot/Istiod, the external controller is just for backward compatibility
- Safely detects cleanly exited pods with no race conditions
- Requires no modification of workload

**Drawbacks**
- Adds another component to an Istio install

---

**Internal Process Inspector**
This option involves activating the Shared Process Namespace feature https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace/).

We add code (via an init or new proxy code) which (after some delay) inspects the process namespace and attempts to determine if it is the only workload running (ignoring the pod sandbox container, which would be PID 1).

**Benefits**
- Requires no modification of workload

**Drawbacks**
- This introduces a race condition. If, for example, you create a new pod with a workload which pulls a large image, the pod will start and the istio-proxy container will start its delay timer, and may terminate itself before the workload has even tried to start.
- Inspecting the process tree may be flaky, since we can't predict every permutation of workloads

---
**Envelope**
We create a new binary which starts an envoy proxy and then starts an arbitrary workload as a child process. When the workload exits, the proxy shuts down.

**Benefits**
- Simplest solution, most reliable

**Drawbacks**
- Requires modification of the workload.

**Comment by howardjohn** [comment]
> External Detection Controller

Alternative to that one is to have a controller that queries watches pods, and if they get into a state where envoy is sticking around and its a cronjob, deletes the pod. Changing the proxy to call this controller is too invasive for a hack that will be fixed by k8s soon(ish)

**Comment by stewartbutler** [comment]
That may cause the Job created by the cronjob to continue to reschedule the pod depending on how it is configured. Maybe delete the corresponding Job?

**Comment by howardjohn** [comment]
We would have to look into the details but I am fairly sure it could be done

**Comment by Stono** [comment]
https://github.com/kubernetes/enhancements/issues/753

Fyi

**Comment by Stono** [comment]
BTW https://github.com/istio/istio/issues/20069 would also help

**Comment by smawson** [comment]
A solution where the agent gets notified when it needs to shut down seems a lot safer, is it really a lot more work? It could just open a bidi stream to the controller, which would send it a notification if the application containers have exited, and it could then ask envoy to terminate. Downside is that it would require changes to the agent, vs. a controller that is deleting pods externally.

Maybe someone should play with the deleting pods thing and see how well it works, and if it doesn't pass muster we can move up to an agent-based solution?

Anyone want to prototype something?

**Comment by howardjohn** [comment]
We can't (easily) allow anything to terminate envoy from outside the pod for security reasons. It seems to make sense to me to either have something external detecting and killing, or the agent detecting and killing itself, but not some mix of the two

**Comment by smawson** [comment]
The proposed design was that it is a trusted service (verified through TLS) and an optional feature. So I think we could make the security case that it is fine. The main issue with agent detecting and killing itself is that it doesn't have as much information as a service does.

So the agent is still detecting and killing itself, but based on information it gets from a trusted service (that information being: please exit if safe to do so).

**Comment by mbrowne** [comment]
I am confused about the recommended workaorunds. It seems that checking for liveness by making a request to `http://127.0.0.1:15020/healthz/ready` can only be done from the sidecar. But the examples seem to suggest that you'd be able to put the command that starts your cron job after this in the same block:

```
args:
 - |
   trap "curl --max-time 2 -s -f -XPOST http://127.0.0.1:15020/quitquitquit" EXIT
   while ! curl -s -f http://127.0.0.1:15020/healthz/ready; do sleep 1; done
   echo "Ready!"
   < your job >
```

The `< your job >` part needs to be run from the main container, not the istio sidecar. I might be missing something very basic about istio, but in any case the container running my cronjob still does not terminate after adding the above.

**Comment by wapdat** [comment]
Hi @mbrowne - FWIW it works for me - the job calling 127.0.0.1:15020 to quitquitquit the sidecar.

**Comment by yuenwah** [comment]
Is it possible to create a job-type pod without injecting a sidecar

**Comment by liamawhite** [comment]
Yes, see https://istio.io/docs/setup/additional-setup/sidecar-injection/#more-control-adding-exceptions

**Comment by mbrowne** [comment]
@yuenwah That's what I ended up doing. Note that you have to put it on the pod template, not the job template or anywhere else, or the annotation won't do anything.  For example:

```
    apiVersion: batch/v1beta1
    kind: CronJob
    metadata:
      ...
    spec:
      ...
      jobTemplate:
        spec:
          template:
            metadata:
              annotations:
                # disable istio on the pod due to this issue:
                # https://github.com/istio/istio/issues/11659
                sidecar.istio.io/inject: "false"
```

**Comment by Stono** [comment]
For anyone else landing here also be aware of https://github.com/istio/istio/issues/24058

**Comment by JohnWillker** [comment]
If you use a cronjob with a Golang app inside. This library tries to solve some of these problems: https://github.com/allisson/go-istio-proxy-wait

**Comment by SuleimanWA** [comment]
> 
> 
> @yuenwah That's what I ended up doing. Note that you have to put it on the pod template, not the job template or anywhere else, or the annotation won't do anything. For example:
> 
> ```
> apiVersion: batch/v1beta1
> kind: CronJob
> metadata:
>   ...
> spec:
>   ...
>   jobTemplate:
>     spec:
>       template:
>         metadata:
>           annotations:
>             # disable istio on the pod due to this issue:
>             # https://github.com/istio/istio/issues/11659
>             sidecar.istio.io/inject: "false"
> ```

This is a good solution if RBAC not enabled, when u enable it, this will fail because it considered out of mesh

**Comment by SuleimanWA** [comment]
> 
> 
> @yuenwah That's what I ended up doing. Note that you have to put it on the pod template, not the job template or anywhere else, or the annotation won't do anything. For example:
> 
> ```
> apiVersion: batch/v1beta1
> kind: CronJob
> metadata:
>   ...
> spec:
>   ...
>   jobTemplate:
>     spec:
>       template:
>         metadata:
>           annotations:
>             # disable istio on the pod due to this issue:
>             # https://github.com/istio/istio/issues/11659
>             sidecar.istio.io/inject: "false"
> ```

This is a good solution if RBAC not enabled, when u enable it, this will fail because it considered out of mesh

**Comment by hobbytp** [comment]
One possible solution for job is call "curl -XPOST http://localhost:15000/quitquitquit" to exist envoyproxy when job task finished.

```yaml
kubectl -n hobby apply -f -<<EOF
apiVersion: batch/v1
kind: Job
metadata:
  name: sleep
spec:
  template:
    spec:
      containers:
      - name: sleep
        image: governmentpaas/curl-ssl
        command: ["/bin/sh","-c"]
        args: [ "/bin/sleep 10; curl -XPOST http://localhost:15000/quitquitquit" ]        
        imagePullPolicy: IfNotPresent
      restartPolicy: Never
  backoffLimit: 4
EOF
```

**Comment by howardjohn** [comment]
One note for folks that cannot change their docker images or use shell, you can do some tricks with shared volumes to mount new binaries into your container. for example:
```
apiVersion: v1
kind: Pod
metadata:
  name: shell
spec:
  restartPolicy: Never
  terminationGracePeriodSeconds: 0
  initContainers:
  - name: scuttle-init
    image: howardjohn/scuttle-shell
    volumeMounts:
    - mountPath: /var/lib/scuttle/bin/
      name: scuttle-bootstrap
    command:
    - cp
    - /scuttle
    - /var/lib/scuttle/bin/scuttle
  containers:
  - name: shell
    image: howardjohn/alpine-shell
    imagePullPolicy: IfNotPresent
    command:
      - /var/lib/scuttle/bin/scuttle
      - /bin/sleep
      - "5"
    env:
    - name: ISTIO_QUIT_API
      value: http://localhost:15020
    - name: ENVOY_ADMIN_API
      value: http://localhost:15000
    volumeMounts:
    - mountPath: /var/lib/scuttle/bin/
      name: scuttle-bootstrap
  volumes:
  - name: scuttle-bootstrap
    emptyDir: {}
```

along with https://istio.io/latest/docs/setup/additional-setup/sidecar-injection/#custom-templates-experimental you can probably make this automagically injected. Main problem is you may not know which `command` to run (if pod only sets `args` and not `command`)

**Comment by Bessonov** [comment]
@hzxuzhonghu is there more information?

**Comment by mmerickel** [comment]
Was there a fix committed somewhere? What does it mean to mark this done?

**Comment by hzxuzhonghu** [comment]
https://github.com/istio/istio/issues/11659#issuecomment-809372980

**Comment by mmerickel** [comment]
Thanks - I’ve been using the quitquitquit hack in my containers for over a year since starting to use istio and it’s quite tedious to maintain and remind team members of the pitfalls when deploying into istio. I’ve been following this issue since most helm charts etc do not ship with workarounds for istio like this and I am hoping istio would come up with a solution to wrap one-off pods so that they can shutdown correctly without manual modifications. I’d hope this issue would stay open.

**Comment by howardjohn** [comment]
I think this should be open still, the above is a workaround. It should be automatic

**Comment by tomfankhaenel** [comment]
This is indeed an issue worth solving within istio itself. One does not simply think of this kind of behavior when deploying istio. I was expecting this could be solved by adding some kind of magic istio lable or annotation.

**Comment by rdavyd** [comment]
Ran into the issue using the workaround with calling `/quitquitquit` endpoint on k8s version 1.22.5. 
POD shows as `Completed`, even when the main job container finishes with error.
![image](https://user-images.githubusercontent.com/57534322/150560563-c259c6f2-2d54-4d02-a785-61fe0cde36ce.png)
Didn't test that on earlier k8s versions.

**Comment by howardjohn** [comment]
> POD shows as Completed, even when the main job container finishes with error.

That is how Jobs work in Kubernetes regardless of Istio. I think there is some config somewhere to tweak it

**Comment by rdavyd** [comment]
@howardjohn Did some digging and testing and did not find a way to do such a tweak. Looks like k8s takes the displayed PODStatus from one of the containers. Tested POD with 2 containers and without `istio-proxy` and its `PODStatus` always returned the exit status of the first container in the array.
But when adding `istio-proxy` (it adds as the second container) this behavior changes and `PODStatus` displays the exit status of `istio-proxy`.
Also didn't find envoy API call like `/quitquitquit` that forces exit with non zero status.

**Comment by sathieu** [comment]
Hello, I've done an "operator" to handle this problem until [keystone containers](https://github.com/kubernetes/enhancements/pull/2869) are added to k8s.

See: https://gitlab.com/kubitus-project/kubitus-pod-cleaner-operator/-/blob/main/README.md

**Comment by ceastman-r7** [comment]
This seems to maintain the error code from the job:

x=$(echo $?); curl -fsI -X POST http://localhost:15020/quitquitquit && exit $x

**Comment by andrewhibbert** [comment]
Seems it is not always the same port. The following works for me

```
x=(echo $?); curl -fsI -X POST http://localhost:15000/quitquitquit && exit $x
```

**Comment by jacekszlachtass** [comment]
dollar sign is missing in the above command:

```
x=$(echo $?); curl -fsI -X POST http://localhost:15000/quitquitquit && exit $x
```

**Comment by kschu91** [comment]
Just if anyone is searching for a simple solution: https://github.com/AOEpeople/kubernetes-sidecar-cleaner

We have developed a simple app to clean up the istio-proxies in completed CronJobs.

**Comment by ZiaUrRehman-GBI** [comment]
@kschu91 where should be this controller install? Inside istio namespace?

**Comment by brunocascio** [comment]
Is there any plan to implement something at istio level rather than using hacky solutions or 3rd party operator?

**Comment by mmerickel** [comment]
https://istio.io/latest/blog/2022/introducing-ambient-mesh/ can't come soon enough. :-)

**Comment by ceastman-r7** [comment]
There is a kubernetes enhancement that should allow orderly startup and shutdown of multi container pods: https://github.com/kubernetes/enhancements/issues/3759

or ambient :)

**Comment by nickzelei** [comment]
> Just if anyone is searching for a simple solution: https://github.com/AOEpeople/kubernetes-sidecar-cleaner
> 
> We have developed a simple app to clean up the istio-proxies in completed CronJobs.

This worked perfectly for me - thank you for your contribution!

@ZiaUrRehman-GBI - you can deploy it anywhere. The helmchart is deployed with a cluster role so it moderated pods in the entire cluster. At least that's how it works with the helmchart that @kschu91 is providing.
You'd have to roll your own deployment with their image if you wanted it to be scoped to a specific namespace.

**Comment by brunocascio** [comment]
So there's no official solution.

I prefer to use a script to notify Istio rather than installing an unofficial or hacky operator 😃 

Thanks everyone

**Comment by bmitchinson** [comment]
Not a solution by any means, but a reminder that you can opt out of istio when appropriate by using the `sidecar.istio.io/inject: "false"` pod annotation.

**Comment by yasensim** [comment]
Sidecar as a first class citizen is in alpha in behind a feature gate in Kubernetes 1.28 and is beta and enabled by default in Kubernetes 1.29. If/when Istio implements restartPolicy the istio-proxy lifecycle will align to the main container. This will allow running Jobs on Istio.
For reference [HERE](https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/) .

cc: @howardjohn

**Comment by howardjohn** [comment]
Istio already supports it behind the ENABLE_NATIVE_SIDECARS. Note I highly recommend only using it on K8s 1.29 as some of the lifecycle is broken in 1.28

**Comment by kmushegi** [comment]
> Just if anyone is searching for a simple solution: https://github.com/AOEpeople/kubernetes-sidecar-cleaner
> 
> We have developed a simple app to clean up the istio-proxies in completed CronJobs.

This is exactly what I needed, thank you!!



> @kschu91 where should be this controller install? Inside istio namespace?

I installed it into the `istio-system` namespace since the watcher watches all namespaces: https://github.com/AOEpeople/kubernetes-sidecar-cleaner/blob/main/main.go#L54 and runs with a clusterrole

**Comment by Stono** [comment]
I'm going to close this supperrrr old issue.
Native sidecars is the solution (albeit some gotchas i've talked about here https://karlstoney.com/moving-to-native-sidecars/)

**Comment by Callek** [comment]
> I'm going to close this supperrrr old issue. Native sidecars is the solution (albeit some gotchas i've talked about here https://karlstoney.com/moving-to-native-sidecars/)

@Stono from that linked article they actually recommend not using it with istio:

> Bit of an update here; I wouldn't do it with Istio just yet! There's a [reasonably nasty bug](https://github.com/istio/istio/issues/51855?ref=karlstoney.com) I've stumbled across which means outbound connections from your app will start seeing Connection: close. This has pretty horrid interactions with some HTTP connection pools.

**Comment by Stono** [comment]
I know, I wrote it :)

My point was more that there won't be a "solution" to this multi year problem, when the _actual_ solution _is_ Sidecar containers.  So the choices are either:

- Continue with hacky script work arounds (which work, i used them for years)
- Move to Sidecar containers, but be aware of those couple of gotchas I mentioned in the blog.  I've detailed my mitigation to the shutdown issue here https://github.com/istio/istio/issues/51855?ref=karlstoney.com#issuecomment-2203894025 which seems to work fine for me, i've been using Sidecar containers in production for a while now.

**Comment by ruslanguns** [comment]
Hi @Stono, thanks for your article! I’m still a bit confused about the best way to set this up in a cron job. Could you provide some guidance on how to properly configure the startupProbe with Istio Proxy? What would be the best approach for this?

**Comment by nikhilbangar8** [comment]
> Use sidecar containers, they start before your app.
> […](#)

Using Kubernetes native Sidecar Containers Helped with issues for me. Thanks @Stono for the blog https://karlstoney.com/moving-to-native-sidecars/
